_content/the-things-stack-v3/migrate-to-v3.md

```diff
@@ -116,6 +120,123 @@ Since you have not migrated your gateway from V2 deployment yet, the new Join re
 
 Interesting thing is that your end device's new Join request will be received by the V3 server too! You can thank for this to [Packet Broker](https://www.thethingsindustries.com/docs/reference/peering/#packet-broker) (you can verify this by observing the uplinks metadata in V3). Now, V3 server will accept this Join request, so your end device will get a new V3 `DevAddr`, channel settings and other MAC parameters. Based on a newly assigned `DevAddr`, Packet Broker will route the traffic to your V3 network.
 
-# What to do with ABP end device?
+## What to do with ABP end device?
+
+If you are using an ABP device, that means your end device is still associated with the V2 `DevAddr`, so the Packet Broker will not be able to route its traffic to your V3 server if you have not connected your gateway to V3. In order to route this traffic correctly, follow the next steps to migrate your gateway to the V3 deployment.
+
+# Method 2: Migrate end devices using the migration tool
+
+To complete the migration this way, you need to have  [`ttn-lw-migrate` tool](https://github.com/TheThingsNetwork/lorawan-stack-migrate) installed. 
+
+`ttn-lw-migrate` is used to export end devices and applications from The Things Network Stack V2 deployment to a [JSON file](https://www.thethingsindustries.com/docs/getting-started/migrating/device-json/). This JSON file can later be [imported](https://www.thethingsindustries.com/docs/getting-started/migrating/import-devices/) in The Things Stack V3 via Console or via CLI.
+
+## Note about migrating an active session
+
+After exporting an end device by using the migration tool,  session and root keys will be cleared from The Things Network Stack V2 deployment by default. This means that these devices will no longer work in the V2 deployment. You can disable this behavior by using the `--ttnv2.with-session=false` flag when running `ttn-lw-migrate`.
+
+If it was the public V2 deployment, and you are using The Things Stack v3.10 or higher, you may transfer the active device session as well - this means your device can continue working on The Things Stack V3 deployment without the need for a rejoin.
+
+If it was the private V2 deployment, devices that have the `DevAddr` that is out of the scope of the address block supported by The Things Stack V3 will have to perform a rejoin. 
```

Comment:
For private V2 SaaS customers, we do make the DevAddr block available to The Things Stack Cloud. So that is going to work, also with ABP.

Reply:
@johanstokking I found this here: https://www.thethingsindustries.com/docs/getting-started/migrating/migrating-from-v2/, so I guess this should be updated then. Btw, something is confusing me a bit, so I would be grateful if you would explain this shortly.

- If you migrate an active OTAA session to V3, keys get deleted from V2 automatically, and the device doesn't have to perform a new join to V3. If you use `--ttnv2.with-session=false` flag, keys will stay untouched on V2, but then you have to manually prevent the device from joining V2 and your device has to do a new join to the V3 network
- What confuses me is 1 session during ABP device lifetime -> what would `--ttnv2.with-session=false` flag mean in ABP case? Just that keys don't get deleted from V2 and you have to manually delete the device from V2 in order for it to work on V3? Is this flag even applicable to ABP?

Reply:
> I found this here: [thethingsindustries.com/docs/getting-started/migrating/migrating-from-v2](https://www.thethingsindustries.com/docs/getting-started/migrating/migrating-from-v2/)

I referred to the second paragraph, about the DevAddr scope. That is not covered in that doc, right?

> * What confuses me is 1 session during ABP device lifetime -> what would `--ttnv2.with-session=false` flag mean in ABP case? Just that keys don't get deleted from V2 and you have to manually delete the device from V2 in order for it to work on V3? Is this flag even applicable to ABP?

Yes. Indeed, not deleting the ABP session in V2 and creating it in V3 leads to errors as the session is then in two places, creating potential conflicts.

Reply:
> > I found this here: [thethingsindustries.com/docs/getting-started/migrating/migrating-from-v2](https://www.thethingsindustries.com/docs/getting-started/migrating/migrating-from-v2/)
> 
> I referred to the second paragraph, about the DevAddr scope. That is not covered in that doc, right?

Actually, it is - second paragraph under **Migrating active sessions**.

Reply:
Ah I see what you mean now, sorry. This:

>When migrating from a private The Things Network Stack V2, devices that are outside of the DevAddr address block supported by The Things Stack Cloud will have to rejoin the network, otherwise The Things Stack will be unable to route their uplink and downlink traffic.

This indeed is not true and can be removed or rewritten as that we are making the same DevAddr range available in The Things Stack Cloud.